iac-extensions/arm/src/main/java/org/sonar/iac/arm/parser/ArmParser.java

```diff
@@ -99,31 +125,113 @@ private static List<ResourceDeclaration> convertResources(SequenceTree resource)
+    Property copyInput = properties.remove("copy.input");
+    Property value = properties.remove("value");
+
+    for (Map.Entry<String, Property> unexpectedProperty : properties.entrySet()) {
+      TextRange position = unexpectedProperty.getValue().textRange();
+      LOG.debug("Unexpected property '{}' found in output declaration at {}:{}, ignoring it.", unexpectedProperty.getKey(), position.start().line(), position.start().lineOffset());
+    }
+
+    checkMandatoryObject(tupleTree.metadata(), "type", type);
+
+    return new OutputDeclarationImpl(name, type, condition, copyCount, copyInput, value);
+  }
+
+  /**
+   * This method will raise a Parse exception if any of the provided object is null.
+   * It require first the metadata object to report the location in the exception.
+   * Then it is expecting arguments by pair: the name of the argument (for the exception message) and the object reference to check.
+   */
+  private static void checkMandatoryObject(YamlTreeMetadata metadata, Object... objects) {
+    List<String> missing = new ArrayList<>();
+    for (int i = 0; i < objects.length - 1; i += 2) {
+      String objectName = (String) objects[i];
+      Object objectReference = objects[i + 1];
+      if (objectReference == null) {
+        missing.add(objectName);
       }
     }
 
-    if (type == null || version == null || name == null) {
-      TextPointer pointer = new BasicTextPointer(tree.metadata().textRange());
-      throw new ParseException(String.format("Resource without required field (name, type, apiVersion) spotted at %d:%d", pointer.line(), pointer.lineOffset()), pointer, null);
+    if (!missing.isEmpty()) {
+      TextPointer pointer = new BasicTextPointer(metadata.textRange());
+      StringBuilder errorMessage = new StringBuilder()
+        .append("Missing required field")
+        .append(missing.size() > 1 ? "s" : "")
+        .append(" [\"").append(StringUtils.join(missing, "\", \"")).append("\"]")
+        .append(" at ").append(pointer.line()).append(":").append(pointer.lineOffset());
+      throw new ParseException(errorMessage.toString(), pointer, null);
     }
+  }
 
-    return new ResourceDeclarationImpl(name, version, type, otherProperties);
+  private static MappingTree toMappingTree(YamlTree tree) {
+    if (!(tree instanceof MappingTree)) {
+      throw new ParseException("Expected MappingTree, got " + tree.getClass().getSimpleName(), new BasicTextPointer(tree.metadata().textRange()), null);
+    }
+    return (MappingTree) tree;
+  }
+
+  /**
+   * Transform a collection of TupleTree into a Map of Property to easily find specific properties and process them.
+   * This is a recursive method which will also convert sub-object, adding a prefix at every level.
+   * Example :
+   * {
+   *   "key1" : "value1",
+   *   "key2 : {
+   *     "key3" : "value3"
+   *   }
+   * }
+   * => Will be translated into this Map {"key1":"value1", "key2.key3":"value3"}
```

Comment:
What if `key2` contains a dot? It's a valid string in JSON. I'm not sure what character might be good here as separator

Reply:
:thinking_face: I didn't think about it.
Maybe something that can be tackled separately in the ticket I created [Improve Property to support array and object](https://sonarsource.atlassian.net/browse/SONARIAC-840).
I will check if I can change the approach quickly here otherwise.